### DISCUSSION TITLE
Ermahgerd, Berks!

### DISCUSSION THREAD

**Discussion by teunbrand** [ggplot2-extenders/ggplot-extension-club#62] [General]
Not Gersberms, but I came across this book via bluesky somewhere: https://ggplot2-extended-book.com/.
Seems like WIP but thought it might be of interest to this audience.
Haven't read parts of it myself yet.

**Comment by EvaMaeRey** [comment]
I thought this was post was spam! Had to do some research on 'Ermahgerd'!  But now I'm up to date.

Looks like an ambitious project.  I guess thematically looking at what extensions aim to achieve - which is not tackled in extension website, for example.  

I would be cool to go through extension functions and make some lists by prefix, e.g. theme_ geom_ facet_...  It seems like extensions website does author, CRAN, package, but not prefix?

**Reply by teunbrand** [reply]
> Had to do some research on 'Ermahgerd'!

You mean you didn't grow up on the finest of early 2010s internet meme culture?

> I would be cool to go through extension functions and make some lists by prefix

You can type queries like the following into GitHub search `org:cran symbol:facet_` and get reasonably decent hits.

> which is not tackled in extension website

You kind of have to volunteer your extension for the website, it is not just a single person keeping their eyes on the CRAN incoming feed and finding out if new packages are extensions. I'd be a hard ask for people to organise the information for their package any further.

**Reply by EvaMaeRey** [reply]
I don't know my memes! 😭 GitHub search search sounds right.

**Reply by EvaMaeRey** [reply]
300+ facet_* hits. I kind of think it would be cool to have an exported functions filter in the extensions website. I think text filter just searches the summary that is submitted, and not the package contents, and I guess exported functions are not typically submitted. 

![image](https://github.com/user-attachments/assets/17d485fd-fb64-4378-ba5f-f0a6f5c35540)

This also seemed like a possibly good task for LLMs, got some invented stuff.  patchwork::facet_patchwork doesn't exist I think, and ggplot2::facet_nested which is ggh4x?

**Reply by EvaMaeRey** [reply]
I think looking at exported function prefixes might give a nice heuristic about which packages use 'gg' prefix to mean ggplot2 object generating versus 'gg' as in grammar grammar ... Along the lines of the dichotomy @corybrunson laid out at the extension ecosystem Hadley meeting. I guess gathering those should be pretty straightforward, especially if constrained to 140 registered extensions.  Get repo urls, collect namespaces. organize.  Done.  network graph!

**Reply by yjunechoe** [reply]
As I have meme-d about previously, my preferred heuristic 😆 :

![image](https://github.com/user-attachments/assets/029d4a03-2b53-441e-813b-0f848e50e4c0)


https://fosstodon.org/@yjunechoe/111120886792143682

**Reply by EvaMaeRey** [reply]
Fun meme! We are staying on the meme topic!  I think the exported function prefixes is more about API, user-experience aspirations than implementation (and durability?).

**Reply by yjunechoe** [reply]
Yeah I see what you mean (and that requires a more careful approach indeed!)

Just to drop a proof of concept:

```r
library(tidyverse)
pkgs <- as_tibble(tools::CRAN_package_db())
gg_pkgs <- pkgs %>% 
  select(Package, Depends, Imports, Suggests) %>% 
  filter(
    str_detect(Package, "^gg"),
    if_any(-Package, ~ str_detect(.x, "ggplot2"))
  ) %>% 
  mutate(
    dep_grid = str_detect(Depends, "\\bgrid\\b") | str_detect(Imports, "\\bgrid\\b"),
    dep_dplyr = str_detect(Depends, "\\bdplyr\\b") | str_detect(Imports, "\\bdplyr\\b")
  ) %>% 
  drop_na()

gg_pkgs %>% 
  split(~ dep_grid + dep_dplyr)
#> $FALSE.FALSE
#> # A tibble: 51 × 6
#>    Package    Depends                            Imports Suggests dep_grid dep_dplyr
#>    <chr>      <chr>                              <chr>   <chr>    <lgl>    <lgl>    
#>  1 ggbeeswarm R (>= 3.5.0), ggplot2 (>= 3.3.0)   beeswa… "gridEx… FALSE    FALSE    
#>  2 ggbiplot   R (>= 3.5.0), ggplot2              scales  "corrpl… FALSE    FALSE    
#>  3 ggbrace    R (>= 4.3)                         ggplot… "knitr,… FALSE    FALSE    
#>  4 ggbuildr   ggplot2                            purrr,… "testth… FALSE    FALSE    
#>  5 ggcorrplot R (>= 3.3), ggplot2 (>= 3.3.6)     reshap… "testth… FALSE    FALSE    
#>  6 ggcorset   R (>= 3.5.0)                       ggplot… "rmarkd… FALSE    FALSE    
#>  7 ggdark     R (>= 3.1)                         ggplot… "testth… FALSE    FALSE    
#>  8 ggdemetra  R (>= 3.1.2), ggplot2 (>= 2.0.0),… ggrepe… "knitr,… FALSE    FALSE    
#>  9 ggdendro   R(>= 3.5)                          MASS, … "rpart … FALSE    FALSE    
#> 10 ggdensity  ggplot2                            isoban… "vdiffr… FALSE    FALSE    
#> # ℹ 41 more rows
#> 
#> $TRUE.FALSE
#> # A tibble: 39 × 6
#>    Package      Depends            Imports               Suggests dep_grid dep_dplyr
#>    <chr>        <chr>              <chr>                 <chr>    <lgl>    <lgl>    
#>  1 ggalign      ggplot2 (>= 3.3.0) "cli, grid, gtable, … "bookdo… TRUE     FALSE    
#>  2 ggallin      ggplot2 (>= 2.2.1) "scales, grid"        "knitr,… TRUE     FALSE    
#>  3 gganimate    ggplot2 (>= 3.5.0) "cli, glue, grDevice… "av, ba… TRUE     FALSE    
#>  4 ggarrow      ggplot2 (>= 3.5.0) "cli, grid, polyclip… "covr, … TRUE     FALSE    
#>  5 ggblend      R (>= 4.2)         "methods, grid, ggpl… "covr, … TRUE     FALSE    
#>  6 ggdaynight   R (>= 2.10)        "ggplot2, grid"       "testth… TRUE     FALSE    
#>  7 ggdist       R (>= 3.6.0)       "ggplot2 (>= 3.5.0),… "tidyse… TRUE     FALSE    
#>  8 ggDoubleHeat R (>= 3.5.0)       "rlang, grid, ggplot… "rmarkd… TRUE     FALSE    
#>  9 ggExtra      R (>= 3.1.0)       "colourpicker (>= 1.… "knitr … TRUE     FALSE    
#> 10 ggfittext    R (>= 3.6)         "grid (>= 3.1), stri… "knitr,… TRUE     FALSE    
#> # ℹ 29 more rows
#> 
#> $FALSE.TRUE
#> # A tibble: 38 × 6
#>    Package     Depends                           Imports Suggests dep_grid dep_dplyr
#>    <chr>       <chr>                             <chr>   <chr>    <lgl>    <lgl>    
#>  1 ggalignment R (>= 3.4)                        "dplyr… "rmarkd… FALSE    TRUE     
#>  2 ggalluvial  R (>= 3.6), ggplot2 (>= 2.2)      "stats… "grid, … FALSE    TRUE     
#>  3 ggautomap   R (>= 4.1),                       "carto… "ggrepe… FALSE    TRUE     
#>  4 ggbrain     R (>= 3.5.0)                      "Matri… "knitr,… FALSE    TRUE     
#>  5 ggBubbles   R (>= 3.5.0)                      "dplyr… "BiocSt… FALSE    TRUE     
#>  6 ggcharts    R (>= 3.5.0), ggplot2 (>= 3.0.0)  "color… "gapmin… FALSE    TRUE     
#>  7 ggcleveland R (>= 3.6.0)                      "dplyr… "testth… FALSE    TRUE     
#>  8 ggdag       R (>= 3.4.0)                      "dagit… "covr, … FALSE    TRUE     
#>  9 ggESDA      ggplot2, R (>= 3.5.0), tidyverse… "rlang… "testth… FALSE    TRUE     
#> 10 ggFishPlots R (>= 3.5.0), ggplot2             "dplyr… "knitr,… FALSE    TRUE     
#> # ℹ 28 more rows
#> 
#> $TRUE.TRUE
#> # A tibble: 18 × 6
#>    Package      Depends                          Imports Suggests dep_grid dep_dplyr
#>    <chr>        <chr>                            <chr>   <chr>    <lgl>    <lgl>    
#>  1 ggalt        R (>= 3.2.0), ggplot2 (>= 2.2.1) "utils… "testth… TRUE     TRUE     
#>  2 ggedit       R (>= 4.3.0), ggplot2 (>= 3.5.0) "plyr,… "testth… TRUE     TRUE     
#>  3 ggfields     R (>= 4.1.0)                     "dplyr… "Copern… TRUE     TRUE     
#>  4 ggfortify    methods, ggplot2 (>= 2.0.0)      "dplyr… "testth… TRUE     TRUE     
#>  5 ggfun        R (>= 4.1.0)                     "cli, … "ggplot… TRUE     TRUE     
#>  6 gggenomes    R (>= 3.4.2), ggplot2 (>= 3.5.0… "vctrs… "testth… TRUE     TRUE     
#>  7 ggiraphExtra R (>= 2.10)                      "ggplo… "TH.dat… TRUE     TRUE     
#>  8 ggmap        R (>= 3.1.0), ggplot2 (>= 2.2.0) "png, … "MASS, … TRUE     TRUE     
#>  9 ggmulti      R (>= 3.4.0), methods, ggplot2   "stats… "png, t… TRUE     TRUE     
#> 10 ggpicrust2   R (>= 3.5.0)                     "ALDEx… "Biobas… TRUE     TRUE     
#> 11 ggpmisc      R (>= 4.0.0), ggpp (>= 0.5.8)    "grid,… "knitr … TRUE     TRUE     
#> 12 ggpol        ggplot2 (>= 3.3.0), R (>= 3.4.0) "grid,… "knitr,… TRUE     TRUE     
#> 13 ggpp         R (>= 4.0.0), ggplot2 (>= 3.5.0) "stats… "knitr … TRUE     TRUE     
#> 14 ggpubr       R (>= 3.1.0), ggplot2 (>= 3.4.0) "ggrep… "grDevi… TRUE     TRUE     
#> 15 ggraph       R (>= 2.10), ggplot2 (>= 3.5.0)  "dplyr… "networ… TRUE     TRUE     
#> 16 ggsankeyfier R (>= 4.1.0)                     "dplyr… "knitr,… TRUE     TRUE     
#> 17 ggsector     ggplot2, grid                    "dplyr… "BiocMa… TRUE     TRUE     
#> 18 ggseg        R (>= 3.3)                       "ggplo… "knitr,… TRUE     TRUE
```

**Reply by corybrunson** [reply]
I like the comment about sorting the extensions by their type of contribution, on the spectrum from extending the specific implementation of the grammar of graphics to simply building viz packages that produce {ggplot2} plots. Something i've noticed about Tidymodels is that the development team likes to work with past and present engine developers to shore up standards that weren't necessarily followed before. (I'll watch out for an example.) The {ggplot2} extension world is too messy for even an aspirational standard to guide them all, but It would be nice to have a taxonomy that might inspire standards for specific subsets of those extensions. For example, those that basically provide new ggprotos could be provided with guidelines on naming conventions, stat–geom pairings, and proper use of internal names like `group` (which {ggalluvial} fails at), for example, while those that provide new syntactic sugar like {patchwork} could be nudged to avoid overlap and otherwise play well with each other. (And it's not like the original authors should be expected to take all this on, but rather to negotiate with others to tailor PRs to address inconsistencies.)

**Reply by EvaMaeRey** [reply]
I like the pipeline! I wish this was a path to getting names of exported functions.  

gg_pkgs %>% 
  pivot_longer(cols = dep_grid:dep_dplyr) %>% 
  filter(value == T) %>% 
  select(Package, name) %>% 
  ggedgelist:::ggedgelist_quick(include_names = T, 
                                layout = "fr")

![image](https://github.com/user-attachments/assets/83422fdf-2b87-4781-babf-10095d61eef9)

ggedgelist:::ggedgelist() and ggedgelist_quick are 'produces a ggplot2 object' type of gg. 🤦‍♀️😬

**Reply by teunbrand** [reply]
I don't think CRAN displays exports publicly.
Somebody with more webscraping experience than me might be able to get the r-universe exports listing (e.g. https://slowkow.r-universe.dev/ggrepel#exportlist).
Alternatively, you could install all the packages and run through `getNamespaceExports(asNamespace("ggrepel"))` for all of them.

**Reply by EvaMaeRey** [reply]
I think packages on github -> github url -> namespace would in theory get a big enough with the urls coming from ext gallery.  

easy_ and some other bespoke prefixes is kind of getting in the way of seeing other stuff, but maybe with a bit more cleaning, other things emerge.
![image](https://github.com/user-attachments/assets/3baf4234-fb8b-4fb2-9bc5-622812eb63b6)

**Reply by EvaMaeRey** [reply]
narrowing to specific base ggplot2 prefixes....

![image](https://github.com/user-attachments/assets/39164a27-5cd4-4f1e-8712-a3a6b080377d)

**Reply by teunbrand** [reply]
I'm not sure I'm 100% understanding this graph. Nodes are packages/functions and edges are 'function is exported from package'? How frequently does a function node have >1 edge and is there a salient example of this?

**Reply by EvaMaeRey** [reply]
@corybrunson 

> Something i've noticed about Tidymodels is that the development team likes to work with past and present engine developers to shore up standards that weren't necessarily followed before. 

Interesting to hear that that there's this active/involved approach.

> (I'll watch out for an example.) 

Would be interested! 

> The {ggplot2} extension world is too messy for even an aspirational standard to guide them all, but It would be nice to have a taxonomy that might inspire standards for specific subsets of those extensions. For example, those that basically provide new ggprotos could be provided with guidelines on naming conventions, stat–geom pairings, and proper use of internal names like `group` (which {ggalluvial} fails at), for example, while those that provide new syntactic sugar like {patchwork} could be nudged to avoid overlap and otherwise play well with each other. (And it's not like the original authors should be expected to take all this on, but rather to negotiate with others to tailor PRs to address inconsistencies.)

It is a super big space and knowing the space seems tricky - success curse!  But in some ways, like theme_ functions across packages are closer neighbors than their within-package non-theme counterparts, for example, so I guess I just think, beyond trying to place entire packages on a spectrum, being able to find function neighbors (from graphical grammar perspective) might be useful...

**Reply by EvaMaeRey** [reply]
> I'm not sure I'm 100% understanding this graph. Nodes are packages/functions and edges are 'function is exported from package'? How frequently does a function node have >1 edge and is there a salient example of this?

So edges drawn between function name and function prefix, as well as edges between function name and repo name.  

So you have who (author/repo) connected to what: (specific function) connected to where (prefix)  - the where being a conceptual in-graph 'where'.   

So a function should have an edge to its prefix and its package.

Here's a messy webscraping attempt... maybe worthy of a tidytuesday contribution if cleaned up!  I never contributed!
https://evamaerey.github.io/mytidytuesday/2024-11-19-gg-prefixes/gg-prefixes.html

**Reply by teunbrand** [reply]
> So edges drawn between function name and function prefix, as well as edges between function name and repo name.

Thanks I needed this as my brain was confused 😅 Glad that I'm not the only one getting nerdsniped into problems!
I still think that requiring the 'gg' prefix in a package name is a bit strict of a requirement. {patchwork} but also my own {legendry} are listed in the extension page, but don't adhere to the gg-prefix.

**Reply by EvaMaeRey** [reply]
I think grabbing gg.+'s that depend on ggplot2 is the convenient thing to do if looking at CRAN's package list, but patchwork and legendary aren't  if the extension page's offerings is the universe.  Perhaps combining those lists gets you close to the extension packages that are beyond the early experimental phase.   Maybe I should ask the ext page's managers for pitfalls on the scrape.

 I really need to add pink and gray to any dplyr v. grid vis to respect @yjunechoe's meme!

**Reply by corybrunson** [reply]
@EvaMaeRey

> Interesting to hear that that there's this active/involved approach.

I still don't have an example handy, but when i last came across this activity there was a link to this WIP: [_Conventions for R Modeling Packages_](https://tidymodels.github.io/model-implementation-principles/).

> It is a super big space and knowing the space seems tricky - success curse! But in some ways, like theme_ functions across packages are closer neighbors than their within-package non-theme counterparts, for example, so I guess I just think, beyond trying to place entire packages on a spectrum, being able to find function neighbors (from graphical grammar perspective) might be useful...

Oh, that's a great point. I imagine that some extensions even provide both new ggproto classes and single-use wrappers.

**Reply by EvaMaeRey** [reply]
> ggedgelist:::ggedgelist() and ggedgelist_quick are 'produces a ggplot2 object' type of gg. 🤦‍♀️😬

So writing this and maybe some of the voronoi edge stuff, #63, got me thinking about edgelist Stats, which could allow you to feed edgelist df right to ggplot() function.

It is pretty 'zero-featured' and proof of concept, but interesting to think about!  I like being able to specify 'from' and 'to' columns as aes.  Simple example: using gapminder, continent-country edges give you five stars. Oceana, Europe, Americas, Africa, Asia.

```{r}
gapminder::gapminder %>% 
  ggplot() + 
  aes(from = continent, to = country) + 
  geom_point(stat = StatEdgelistNodes) + 
  geom_segment(stat = StatEdgelistEdges)
```

![image](https://github.com/user-attachments/assets/c4057054-1d4c-4565-9460-63e428229d4a)

lots of ggraph and tidygraph igraph under the hood:  https://evamaerey.github.io/mytidytuesday/2024-11-21-ggedgelist-rewrite/ggedgelist-rewrite.html

A tad awkward/inefficient is that layout is computed for each layer.

**Reply by EvaMaeRey** [reply]
Revisiting the Namespace webscraping to get exported functions, get this csv:

https://raw.githubusercontent.com/EvaMaeRey/mytidytuesday/refs/heads/main/2024-11-19-gg-prefixes/exported_funs_exts_ggplot2_tidyverse_org.csv    - kind of fun to browse!

![image](https://github.com/user-attachments/assets/a1c3d708-db99-4c6d-872e-ab64b205c449) 

https://evamaerey.github.io/mytidytuesday/2024-11-19-gg-prefixes/ext_ggplot2_functions_cleaning.html

Still not 100%, but I think it's closer than before... main and master were both NAMESPACE addresses that are in the mix, but I only grabbed main, I think in the first round.

**Reply by EvaMaeRey** [reply]
Some initial exploration (but not looking at linkages anymore, so just who, package, and prefixes)..

``` r
library(tidyverse)
user_repo_fun <- read_csv("https://raw.githubusercontent.com/EvaMaeRey/mytidytuesday/refs/heads/main/2024-11-19-gg-prefixes/exported_funs_exts_ggplot2_tidyverse_org.csv")

user_repo_fun %>% 
  ggplot() + 
  aes(id = repo) + 
  ggcirclepack::geom_circlepack() + 
  ggcirclepack::geom_circlepack_text(color = "whitesmoke") + 
  aes(size = after_stat(1/nchar(id)*area)) +
  ggstamp::theme_void_fill("midnightblue") + 
  coord_equal()
```

![](https://i.imgur.com/Hyz0Ke3.png)

``` r

last_plot() + 
  aes(id = user)

```

![](https://i.imgur.com/s73ySSZ.png)

``` r

last_plot() + 
  aes(id = fun_exported %>% str_extract(".*?_"))

```

![](https://i.imgur.com/QopuaTA.png)

``` r

last_plot() + 
  aes(id = fun_exported %>% str_extract(".+_"))

```

![](https://i.imgur.com/KABqWbl.png)

<sup>Created on 2024-12-06 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>
